library/psa_crypto.c

```diff
@@ -4695,6 +4696,179 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
 }
 #endif /* MBEDTLS_PSA_BUILTIN_KEY_TYPE_DES */
 
+/*
+* ECC keys on a Weierstrass elliptic curve require the generation
+* of a private key which is an integer
+* in the range [1, N - 1], where N is the boundary of the private key domain:
+* N is the prime p for Diffie-Hellman, or the order of the
+* curve’s base point for ECC.
+*
+* Let m be the bit size of N, such that 2^m > N >= 2^(m-1).
+* This function generates the private key using the following process:
+*
+* 1. Draw a byte string of length ceiling(m/8) bytes.
+* 2. If m is not a multiple of 8, set the most significant
+*    (8 * ceiling(m/8) - m) bits of the first byte in the string to zero.
+* 3. Convert the string to integer k by decoding it as a big-endian byte string.
+* 4. If k > N - 2, discard the result and return to step 1.
+* 5. Output k + 1 as the private key.
+*
+* This method allows compliance to NIST standards, specifically the methods titled
+* Key-Pair Generation by Testing Candidates in the following publications:
+* - NIST Special Publication 800-56A: Recommendation for Pair-Wise Key-Establishment
+*   Schemes Using Discrete Logarithm Cryptography [SP800-56A] §5.6.1.1.4 for
+*   Diffie-Hellman keys.
+*
+* - [SP800-56A] §5.6.1.2.2 or FIPS Publication 186-4: Digital Signature
+*   Standard (DSS) [FIPS186-4] §B.4.2 for elliptic curve keys.
+*/
+#if defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR) || \
+    defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY) || \
+    defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR) || \
+    defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_PUBLIC_KEY)
+static psa_status_t psa_generate_derived_ecc_key_weierstrass_helper(
+    psa_key_slot_t *slot,
+    size_t bits,
+    psa_key_derivation_operation_t *operation,
+    uint8_t **data,
+    unsigned *error)
+{
+    mbedtls_mpi N;
+    mbedtls_mpi k;
+    mbedtls_mpi diff_N_2;
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    psa_status_t status;
+
+    mbedtls_mpi_init( &k );
+    mbedtls_mpi_init( &N );
+    mbedtls_mpi_init( &diff_N_2 );
+
+    psa_ecc_family_t curve = PSA_KEY_TYPE_ECC_GET_FAMILY(
+                                slot->attr.type );
+    mbedtls_ecp_group_id grp_id =
+        mbedtls_ecc_group_of_psa( curve, bits, 0 );
+
+    mbedtls_ecp_group ecp_group;
+    mbedtls_ecp_group_init( &ecp_group );
+
+    if( ( status = mbedtls_ecp_group_load( &ecp_group, grp_id ) ) != 0 )
+    {
+        ret = status;
+        goto cleanup;
+    }
+
+    /* N is the boundary of the private key domain. */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_copy( &N, &ecp_group.N ) );
+    /* Let m be the bit size of N. */
+    size_t m = ecp_group.nbits;
+
+    size_t m_bytes = PSA_BITS_TO_BYTES( m );
+    if (*data == NULL)
+        *data = mbedtls_calloc( 1, m_bytes );
```

Comment:
This function determinates the required size of the buffer and allocates the buffer. This buffer is then used by the caller and released by the caller in positive scenario. 
In case of error path buffer is released by this function and pointer set to `NULL`. Caller will in this case free null pointer which is ok according to C standard (`free(NULL)` has no effect). 
The conditional allocation is done because this function can be called second time when value is out of range. In this case we want to use buffer that is already allocated and just fill it with the new data.

Reply:
Now that `psa_generate_derived_ecc_key_weierstrass_helper` is not called in a loop, `*data` is always `NULL` at this point, right? So now it would make sense for `psa_generate_derived_ecc_key_weierstrass_helper` to allocate unconditionally, because that's the place where you know the size, and it makes sense to use a heap buffer because that buffer will end up in the key store.

This part looks unfinished, that's why I thought you were still working on this PR and I didn't re-review it. Do you think there's still a reason for the `if (*data == NULL)` check? At a cursory glance (I have NOT actually reviewed yet), you could just remove this line and document that the function sets `*data` to memory that it allocates.

Actually, please document the behavior of the function regarding `*data` in any case.

Reply:
Yes, `psa_generate_derived_ecc_key_weierstrass_helper` is now always called with `*data == NULL`, so `(*data == NULL)` is redundant. I will remove it and add comment.

I still didn't address previous comment with request to move `data` buffer to stack. It didn't work for me and I end up with the following failure:
`PSA key derivation: largest possible key .......................... *** stack smashing detected ***: terminated`.
I suspect that local data buffer of size `PSA_BITS_TO_BYTES( PSA_VENDOR_ECC_MAX_CURVE_BITS )` is ok for ECC case but it is not enough for other key types.